internal/backend/rclone/backend.go

```diff
@@ -0,0 +1,278 @@
+package rclone
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"math/rand"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/restic/restic/internal/backend"
+	"github.com/restic/restic/internal/backend/rest"
+	"github.com/restic/restic/internal/debug"
+	"github.com/restic/restic/internal/errors"
+	"golang.org/x/net/context/ctxhttp"
+	"golang.org/x/net/http2"
+)
+
+// Backend is used to access data stored somewhere via rclone.
+type Backend struct {
+	*rest.Backend
+	tr         *http2.Transport
+	cmd        *exec.Cmd
+	waitCh     <-chan struct{}
+	waitResult error
+	wg         *sync.WaitGroup
+	conn       *StdioConn
+}
+
+// run starts command with args and initializes the StdioConn.
+func run(command string, args ...string) (*StdioConn, *exec.Cmd, *sync.WaitGroup, func() error, error) {
+	cmd := exec.Command(command, args...)
+
+	p, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	var wg sync.WaitGroup
+
+	// start goroutine to add a prefix to all messages printed by to stderr by rclone
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		sc := bufio.NewScanner(p)
+		for sc.Scan() {
+			fmt.Fprintf(os.Stderr, "rclone: %v\n", sc.Text())
+		}
+	}()
+
+	r, stdin, err := os.Pipe()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	stdout, w, err := os.Pipe()
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	cmd.Stdin = r
+	cmd.Stdout = w
+
+	bg, err := backend.StartForeground(cmd)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	c := &StdioConn{
+		stdin:  stdout,
+		stdout: stdin,
+		cmd:    cmd,
+	}
+
+	return c, cmd, &wg, bg, nil
+}
+
+// New initializes a Backend and starts the process.
+func New(cfg Config) (*Backend, error) {
+	var (
+		args []string
+		err  error
+	)
+
+	// build program args, start with the program
+	if cfg.Program != "" {
+		a, err := backend.SplitShellStrings(cfg.Program)
+		if err != nil {
+			return nil, err
+		}
+		args = append(args, a...)
+	} else {
+		args = append(args, "rclone")
+	}
+
+	// then add the arguments
+	if cfg.Args != "" {
+		a, err := backend.SplitShellStrings(cfg.Args)
+		if err != nil {
+			return nil, err
+		}
+
+		args = append(args, a...)
+	} else {
+		args = append(args, "serve", "restic", "--stdio")
```

Comment:
It would be nice to have an higher level option that add arguments to rclone without having to write "serve restic --stdio" everytime

Reply:
Note that you can [add options to rclone as env vars](https://rclone.org/docs/#environment-variables) which you might find easier.

Reply:
I just figured this out after writing the review

Reply:
We've added a section to the manual for the `rclone` backend, it also contains the hint.